Use single VAO for shaders -- the attribute bind locations for a_position, a_text_coord and a_normal are set to 0, 1 and 2. This allows us to use the same VAO for lit and unlit shaders in Texture Shader.

Vittalbabu opened the pull request.

thomasflynn commented on GVRf/Framework/jni/objects/mesh.cpp:
```diff
@@ -246,17 +247,15 @@ const float *Mesh::getBoundingSphereInfo() {
-void Mesh::generateVAO(Material::ShaderType key) {
+void Mesh::generateVAO() {
 #if _GVRF_USE_GLES3_
+
```
formatting is off.  maybe you have tabs rather than spaces?

http://gearvrf.org/bin/view/GearVRF/GearVRfWikiGetInvolved#CodeGuide

thomasflynn commented on GVRf/Framework/jni/objects/mesh.h:
```diff
@@ -298,22 +275,21 @@ class Mesh: public HybridObject {
     void setVertexAttribLocV2(GLuint location, std::string key) {
         attribute_vec2_keys_[location] = key;
-    }
+        }
```
formatting

thomasflynn commented:
@rongguodong, did you want to look this over?

rongguodong commented:
I can take a look this weekend.

Date: Thu, 27 Aug 2015 18:59:27 -0700
From: notifications@github.com
To: GearVRf@noreply.github.com
CC: rongguodong@hotmail.com
Subject: Re: [GearVRf] Use single VAO for shaders -- the attribute bind locations for a_position, a_text_coord and a_normal are set to 0, 1 and 2. This allows us to use the same VAO for lit and unlit shaders in Texture Shader. (#236)

@rongguodong, did you want to look this over?


—
Reply to this email directly or view it on GitHub.

chenchao1407 commented:
Seems manual rebase needed for this one. Can you help fix it? @Vittalbabu Thanks in advance.

rongguodong commented:
Yes, I also find conflict to apply this patch. Please rebase with the top of tree.

Also, I tested latest CubeMap on Nexus 6, I see black screen with lots of glErrors in CubemapShader::render. Is it a bug?

danke-sra commented:
@rongguodong 

We tested Note4 and S6 and the lastest cubemap (pull request) are working fine. Can you please paste the nexus 6 log containing glErrors? How did you test the cubemap?

danke-sra commented:
@rongguodong  Hi Guodong, I just remembered one thing. Sometimes, there is some issues when app is starting up. In this case, the screen will be blank. However, if you try to run the app after closing it. The issue may go away. Though I am not 100% sure, this could be what you observed. Can you please try again and let us know if the cubemap works on nexus 6? We will certainly find out why the app doesn't start up properly sometimes. Thanks for your feedback!

rongguodong commented:
@danke-sra I just sync the latest master branch and run cubemap sample, nothing special. I tested other samples like gvrsample and gvrvideo, all of them works fine on my Nexus 6. I think some recent modification of the cubemap shader generate error and is caught by the glCheckError at the end of CubemapShader::render.

I can capture logcat when I have time (maybe tonight). I remember I just see glerror print out every frame. The app does not crash. When I start it, I can see the GVRF logo correctly. When it reached to main scene (when the cubemap shader is executed) the error comes out.

In the meanwhile, maybe you can find a nexus phone to test on it?

Vittalbabu commented:
@rongguodong
I have rebased and fixed the tabs/spaces issues.

rongguodong commented:
Since we always bind position, tex_coord, and normal to attribute location 0, 1, and 2, we do not need glGetAttribLocation in shaders any more.

Also, it is better to use meaningful enums (e.g. POSITION_ATTRIBUTE_LOCATION, TEXCOORD_ATTRIBUT_LOCATION, NORMAL_ATTRIBUTE_LOCATION) other than 0, 1, 2.

Vittalbabu commented:
I will add enums for the 3 numbers.

I will also remove the glGetAttribLocation calls for these 3 attributes.

Vittalbabu commented:
Added the enums for the 3 bind locations. Removed all set position/normal/texcoords calls for all the shaders. The Mesh::generateVAO() functions sets these via the enums.

Vittalbabu commented:
This is defensive programming -- to avoid accidentally doing something like this.

        glEnableVertexAttribArray(GLProgram::POSITION_ATTRIBUTE_LOCATION);
        glVertexAttribPointer(GLProgram::NORMAL_ATTRIBUTE_LOCATION, 3, GL_FLOAT, 0, 0, 0);

rongguodong commented:
@Vittalbabu OK. I am fine with it then.

The code looks fine to me although I have not tested it. If some others tested to verify it works well, I am good to go.

Vittalbabu commented:
I ran  a few of the samples and they are fine. modelviwer, litshader, cubemap, assimpmodelloader.

There is one minor change I will push through -- noticed that I had duplicate code in TextureShader. I will check that into my branch soon.

-    if (use_light) {
-        mesh->generateVAO();
-    } else {
-        mesh->generateVAO();
-    }
+
+    mesh->generateVAO();

chenchao1407 commented:
Verified. Everything goes fine. And please fix the spacing format in mesh.cpp. After that everything should be good to go.

chenchao1407 commented on GVRf/Framework/jni/objects/mesh.cpp:
```diff
@@ -176,17 +177,15 @@ void Mesh::getTransformedBoundingBoxInfo(glm::mat4 *Mat,
 #if _GVRF_USE_GLES3_
+
+	if (vaoInitiliased_)
+		return;
```
These two lines are tab based formatting not space based. 
Please in Window -> Preference -> C/C++ -> Code Style
And then edit the K&R[Space] profile, change the profile name and for Tab policy, change it from tabs only to spaces only.

Then you do Ctrl + Shift + F, you will see that there is something changed.

Thanks in advance.

Vittalbabu replied:
Thanks. Didn’t notice this in the diffs.

I will fix and push it soon (along with another Mesh related change).

From: chenchao1407 [mailto:notifications@github.com]
Sent: Thursday, September 03, 2015 2:19 PM
To: Samsung/GearVRf
Cc: Vittalbabu Srinivasan
Subject: Re: [GearVRf] Use single VAO for shaders -- the attribute bind locations for a_position, a_text_coord and a_normal are set to 0, 1 and 2. This allows us to use the same VAO for lit and unlit shaders in Texture Shader. (#236)


In GVRf/Framework/jni/objects/mesh.cpp<https://github.com/Samsung/GearVRf/pull/236#discussion_r38697951>:

>  #if _GVRF_USE_GLES3_

> +

> +     if (vaoInitiliased_)

> +            return;

These two lines are tab based formatting not space based.
Please in Window -> Preference -> C/C++ -> Code Style
And then edit the K&R[Space] profile, change the profile name and for Tab policy, change it from tabs only to spaces only.

Then you do Ctrl + Shift + F, you will see that there is something changed.

Thanks in advance.

—
Reply to this email directly or view it on GitHub<https://github.com/Samsung/GearVRf/pull/236/files#r38697951>.

Vittalbabu commented:
@chenchao1407 
Replaced the tabs by spaces.

chenchao1407 commented:
Hi @Vittalbabu 

You just said that you would push another change to it. Do you want it to be merged now or after that?

Vittalbabu commented:
@chenchao1407 

I need to test that change a little bit more (it complies and the samples I have tested run fine). Let's push this change out to parent and I will submit a new pull request.

Thanks.
